ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatSend.java

```diff
@@ -110,44 +111,47 @@ private void calculateOffsets(boolean enableDataPrefetch) throws IOException, Me
 
-          headerVersion.flip();
-          short version = headerVersion.getShort();
+          startTime = SystemTime.getInstance().milliseconds();
+          byte[] headerVersionBytes = new byte[Version_Field_Size_In_Bytes];
+          bufferedInputStream.read(headerVersionBytes, 0, Version_Field_Size_In_Bytes);
+          short version = ByteBuffer.wrap(headerVersionBytes).getShort();
           if (!isValidHeaderVersion(version)) {
             throw new MessageFormatException(
                 "Version not known while reading message - version " + version + ", StoreKey " + readSet.getKeyAt(i),
                 MessageFormatErrorCodes.Unknown_Format_Version);
           }
-          ByteBuffer header = ByteBuffer.allocate(getHeaderSizeForVersion(version));
-          // read the header
+          logger.trace("Calculate offsets, read and verify header version time: {}",
+              SystemTime.getInstance().milliseconds() - startTime);
+
           startTime = SystemTime.getInstance().milliseconds();
-          headerVersion.clear();
-          header.putShort(headerVersion.getShort());
-          readSet.writeTo(i, Channels.newChannel(new ByteBufferOutputStream(header)), Version_Field_Size_In_Bytes,
-              header.capacity() - Version_Field_Size_In_Bytes);
+          byte[] headerBytes = new byte[getHeaderSizeForVersion(version)];
+          bufferedInputStream.read(headerBytes, Version_Field_Size_In_Bytes,
+              headerBytes.length - Version_Field_Size_In_Bytes);
           logger.trace("Calculate offsets, read header time: {}", SystemTime.getInstance().milliseconds() - startTime);
 
           startTime = SystemTime.getInstance().milliseconds();
-          header.flip();
+          ByteBuffer header = ByteBuffer.wrap(headerBytes);
+          header.putShort(version);
+          header.clear();
           MessageHeader_Format headerFormat = getMessageHeader(version, header);
           headerFormat.verifyHeader();
-          int storeKeyRelativeOffset = header.capacity();
+          logger.trace("Calculate offsets, verify header time: {}",
+              SystemTime.getInstance().milliseconds() - startTime);
 
-          StoreKey storeKey = storeKeyFactory.getStoreKey(
-              new DataInputStream(new MessageReadSetIndexInputStream(readSet, i, storeKeyRelativeOffset)));
+          startTime = SystemTime.getInstance().milliseconds();
+          StoreKey storeKey = storeKeyFactory.getStoreKey(new DataInputStream(bufferedInputStream));
           if (storeKey.compareTo(readSet.getKeyAt(i)) != 0) {
             throw new MessageFormatException(
                 "Id mismatch between metadata and store - metadataId " + readSet.getKeyAt(i) + " storeId " + storeKey,
                 MessageFormatErrorCodes.Store_Key_Id_MisMatch);
           }
-          logger.trace("Calculate offsets, verify header time: {}",
+          logger.trace("Calculate offsets, read and verify storeKey time: {}",
```

Comment:
minor: don't know how meaningful these messages are now, since the "read time" might get accounted for elsewhere.

Reply:
It's mainly for debug. Will reduce `logger.trace` for this method.